### ISSUE TITLE
ERROR: MethodError after Pkg.update() to  Optim 0.15.0

### ISSUE THREAD

**Issue by programistawpf** [JuliaNLSolvers/Optim.jl#621]
`julia> result1 = optimize(restmp_a, abc, lower, upper, Fminbox{GradientDescent}(); iterations = 20);
ERROR: MethodError: no method matching Optim.Fminbox{Optim.GradientDescent,T,P} where P where T()

julia> Pkg.status("Optim")
 - Optim                         0.15.0

julia> abc
3-element Array{Float64,1}:
0.5
0.5
0.5
julia> lower
2-element Array{Float64,1}:
1.25
-2.1
julia> upper
2-element Array{Float64,1}:
Inf
Inf
`

**Comment by programistawpf** [comment]
Thankas but not works , other  idea?

julia> result1 = optimize(restmp_a, abc, lower, upper, 
Fminbox(GradientDescent()); iterations = 20)

ERROR: MethodError: no method matching 
optimize(::NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{F
loat64,1}}, ::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,1}, 
::Optim.Fminbox{Optim.GradientDescent{LineS
earches.InitialPrevious{Float64},LineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Void,Optim.##16#18},Float64,O
ptim.##88#90}; iterations=20)

But something new :

julia> result1 = optimize(restmp_a, abc, lower, upper, 
Fminbox(GradientDescent()))
ERROR: Initial x[(1,)]=1.0 is outside of [0.5, 0.0]
Stacktrace:
  [1] 
optimize(::NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}, 
::Array{Float64,1}, ::
Array{Float64,1}, ::Array{Float64,1}, 
::Optim.Fminbox{Optim.GradientDescent{LineSearches.InitialPrevious{Float64},L
ineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Void,Optim.##16#18},Float64,Optim.##88#90}, 
::Optim.Options{Flo
at64,Void}) at 
C:\Users\PC\.julia\v0.6\Optim\src\multivariate/solvers/constrained\fminbox.jl:195
  [2] 
optimize(::NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}, 
::Array{Float64,1}, ::
Array{Float64,1}, ::Array{Float64,1}, 
::Optim.Fminbox{Optim.GradientDescent{LineSearches.InitialPrevious{Float64},L
ineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Void,Optim.##16#18},Float64,Optim.##88#90}) 
at C:\Users\PC\.jul
ia\v0.6\Optim\src\multivariate/solvers/constrained\fminbox.jl:164

julia> lower
3-element Array{Float64,1}:
  0.0
  0.0
  0.0

julia> upper
3-element Array{Float64,1}:
  1.0
  1.0
  1.5

julia> abc
3-element Array{Float64,1}:
  0.5
  0.5
  0.5

Paul

W dniu 2018-06-16 o 22:22, Patrick Kofod Mogensen pisze:

**Comment by pkofod** [comment]
sorry, I only noticed the method part. See my updated answer

**Comment by programistawpf** [comment]
It is working ,|BUT WITHOUT INTERIATION LIMIT|.

|optimize(function(x), low, high, start_points, ||Fminbox(method(); BUT WITHOUT INTERIATION LIMIT ) |
||optimize(f, lower, upper, abc, Fminbox(GradientDescent()); iterations = 
20)| # No working |
|||optimize(f, lower, upper, abc, Fminbox(GradientDescent()))|| # W|||orking| Paul |

W dniu 2018-06-17 o 22:13, Patrick Kofod Mogensen pisze:

**Comment by pkofod** [comment]
Right, I'm sorry for the confusion. If you look at the link @anriseth provided, you'll see that it's

```
optimize(f, lower, upper, abc, Fminbox(GradientDescent()), Optim.Options(iterations = 20))
```

**Comment by programistawpf** [comment]
`
julia> results = optimize(f, lower, upper, initial_x, 
Fminbox(inner_optimizer), Optim.Options(iterations = 20))
ERROR: MethodError: no method matching optimize(::#f, 
::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,1}, :
:Optim.Fminbox{Optim.GradientDescent{LineSearches.InitialPrevious{Float64},LineSearches.HagerZhang{Float64,Base.Ref
Value{Bool}},Void,Optim.##16#18},Float64,Optim.##88#90}, 
::Optim.Options{Float64,Void})
Closest candidates are:
   optimize(::NLSolversBase.OnceDifferentiable, 
::AbstractArray{T<:AbstractFloat,N} where N, ::AbstractArray{T<:Abst
ractFloat,N} where N, ::AbstractArray{T<:AbstractFloat,N} where N, 
::Optim.Fminbox, ::Any) where T<:AbstractFloat a
t 
C:\Users\PC\.julia\v0.6\Optim\src\multivariate/solvers/constrained/fminbox.jl:164
   optimize(::Any, ::AbstractArray{T<:AbstractFloat,N} where N, 
::AbstractArray{T<:AbstractFloat,N} where N, ::Abstr
actArray{T<:AbstractFloat,N} where N, ::Optim.Fminbox; inplace, 
autodiff) where T<:AbstractFloat at C:\Users\PC\.ju
lia\v0.6\Optim\src\multivariate/solvers/constrained/fminbox.jl:151
   optimize(::Any, ::AbstractArray, ::AbstractArray, ::AbstractArray, 
::Optim.SAMIN, ::Optim.Options) at C:\Users\PC
\.julia\v0.6\Optim\src\multivariate/solvers/constrained/samin.jl:60
   ...

julia> optimize(f, lower, upper, initial_x, Fminbox(inner_optimizer))
Results of Optimization Algorithm
  * Algorithm: Fminbox with Gradient Descent
  * Starting Point: [2.0,2.0]
  * Minimizer: [1.2500000000000002,1.5626480963230658]
  * Minimum: 6.250219e-02
  * Iterations: 11
  * Convergence: true
    * |x - x'| ? 0.0e+00: false
      |x - x'| = 2.22e-16
    * |f(x) - f(x')| ? 0.0e+00 |f(x)|: true
      |f(x) - f(x')| = 0.00e+00 |f(x)|
    * |g(x)| ? 1.0e-08: false
      |g(x)| = 4.26e-01
    * Stopped by an increasing objective: false
    * Reached Maximum Number of Iterations: false
  * Objective Calls: 486839
  * Gradient Calls: 486839
`
W dniu 2018-06-18 o 09:38, Patrick Kofod Mogensen pisze:

**Comment by anriseth** [comment]
@programistawpf can you please use the Github interface with code blocks whenever you include code? The formatting from your emails are not easy to read.

**Comment by programistawpf** [comment]
Les more ;)

`
julia> results = optimize(f, lower, upper, initial_x, 
Fminbox(inner_optimizer), Optim.Options(iterations = 20))
ERROR: MethodError: no method matching optimize(::#f, 
::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,1}, :
:Optim.Fminbox{Optim.GradientDescent{LineSearches.InitialPrevious{Float64},LineSearches.HagerZhang{Float64,Base.Ref
Value{Bool}},Void,Optim.##16#18},Float64,Optim.##88#90}, 
::Optim.Options{Float64,Void})
Closest candidates are:
   optimize(::NLSolversBase.OnceDifferentiable, 
::AbstractArray{T<:AbstractFloat,N} where N, ::AbstractArray{T<:Abst
ractFloat,N} where N, ::AbstractArray{T<:AbstractFloat,N} where N, 
::Optim.Fminbox, ::Any) where T<:AbstractFloat a
t 
C:\Users\PC\.julia\v0.6\Optim\src\multivariate/solvers/constrained/fminbox.jl:164
   optimize(::Any, ::AbstractArray{T<:AbstractFloat,N} where N, 
::AbstractArray{T<:AbstractFloat,N} where N, ::Abstr
actArray{T<:AbstractFloat,N} where N, ::Optim.Fminbox; inplace, 
autodiff) where T<:AbstractFloat at C:\Users\PC\.ju
lia\v0.6\Optim\src\multivariate/solvers/constrained/fminbox.jl:151
   optimize(::Any, ::AbstractArray, ::AbstractArray, ::AbstractArray, 
::Optim.SAMIN, ::Optim.Options) at C:\Users\PC
\.julia\v0.6\Optim\src\multivariate/solvers/constrained/samin.jl:60
   ...

julia> optimize(f, lower, upper, initial_x, Fminbox(inner_optimizer))
Results of Optimization Algorithm
  * Algorithm: Fminbox with Gradient Descent
  * Starting Point: [2.0,2.0]
  * Minimizer: [1.2500000000000002,1.5626480963230658]
  * Minimum: 6.250219e-02
  * Iterations: 11
  * Convergence: true
    * |x - x'| ? 0.0e+00: false
      |x - x'| = 2.22e-16
    * |f(x) - f(x')| ? 0.0e+00 |f(x)|: true
      |f(x) - f(x')| = 0.00e+00 |f(x)|
    * |g(x)| ? 1.0e-08: false
      |g(x)| = 4.26e-01
    * Stopped by an increasing objective: false
    * Reached Maximum Number of Iterations: false
  * Objective Calls: 486839
  * Gradient Calls: 486839
`

W dniu 2018-06-18 o 09:38, Patrick Kofod Mogensen pisze:

**Comment by programistawpf** [comment]
`julia> results = optimize(f, lower, upper, initial_x, Fminbox(inner_optimizer), Optim.Options(iterations = 20)) ERROR: MethodError: no method matching optimize(::#f, ::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,1}, : :Optim.Fminbox{Optim.GradientDescent{LineSearches.InitialPrevious{Float64},LineSearches.HagerZhang{Float64,Base.Ref Value{Bool}},Void,Optim.##16#18},Float64,Optim.##88#90}, ::Optim.Options{Float64,Void}) Closest candidates are:   optimize(::NLSolversBase.OnceDifferentiable, ::AbstractArray{T<:AbstractFloat,N} where N, ::AbstractArray{T<:Abst ractFloat,N} where N, ::AbstractArray{T<:AbstractFloat,N} where N, ::Optim.Fminbox, ::Any) where T<:AbstractFloat a t C:\Users\PC\.julia\v0.6\Optim\src\multivariate/solvers/constrained/fminbox.jl:164   optimize(::Any, ::AbstractArray{T<:AbstractFloat,N} where N, ::AbstractArray{T<:AbstractFloat,N} where N, ::Abstr actArray{T<:AbstractFloat,N} where N, ::Optim.Fminbox; inplace, autodiff) where T<:AbstractFloat at C:\Users\PC\.ju lia\v0.6\Optim\src\multivariate/solvers/constrained/fminbox.jl:151   optimize(::Any, ::AbstractArray, ::AbstractArray, ::AbstractArray, ::Optim.SAMIN, ::Optim.Options) at C:\Users\PC \.julia\v0.6\Optim\src\multivariate/solvers/constrained/samin.jl:60   ... julia> optimize(f, lower, upper, initial_x, Fminbox(inner_optimizer)) Results of Optimization Algorithm  * Algorithm: Fminbox with Gradient Descent  * Starting Point: [2.0,2.0]  * Minimizer: [1.2500000000000002,1.5626480963230658]  * Minimum: 6.250219e-02  * Iterations: 11  * Convergence: true    * |x - x'| ? 0.0e+00: false      |x - x'| = 2.22e-16    * |f(x) - f(x')| ? 0.0e+00 |f(x)|: true      |f(x) - f(x')| = 0.00e+00 |f(x)|    * |g(x)| ? 1.0e-08: false      |g(x)| = 4.26e-01    * Stopped by an increasing objective: false    * Reached Maximum Number of Iterations: false  * Objective Calls: 486839  * Gradient Calls: 486839`

**Comment by anriseth** [comment]
@programistawpf I mean that you need to go on github.com and use the interface on the Github webpage, rather than sending your replies via your email system. In the previous posts you can just click the edit button on github.com and fix it there.

Anyway, the MethodError problem is a bug in Optim.
@pkofod I think the `optimize` interface where you specify options only works when you supply a OnceDifferentiable, but not when it dispatches to other functions.
```julia
df = OnceDifferentiable(f...)
optimize(df, lower, upper, x0, Fminbox(inner_optimizer), Optim.Options()) # works
optimize(f, g!, lower, upper, x0, Fminbox(inner_optimizer), Optim.Options()) # does not work
```

**Comment by pkofod** [comment]
ah yes, right.. that's why I opened the Fminbox API issue the other day (too busy,  I keep forgetting things). We should (I should) patch that

**Comment by pkofod** [comment]
This should be fixed in v0.15.1, so try `Pkg.update()` and try again.
